common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.cpp

```diff
@@ -507,6 +552,154 @@ bool base_wlan_hal_dwpal::attach_ctrl_interface(int vap_id)
+        LOG(ERROR) << "Invalid WPA Control socket used for nl events (m_dwpal_ctx == nullptr)";
+        return false;
+    }
+
+    // Passing a lambda with capture is not supported for standard C function
+    // pointers. As a workaround, we create a static (but thread local) wrapper
+    // function that calls the capturing lambda function.
+    static __thread std::function<DWPAL_Ret(struct nl_msg * msg)> nl_handler_cb_wrapper;
+    nl_handler_cb_wrapper = [&](struct nl_msg *msg) -> DWPAL_Ret {
+        if (process_dwpal_nl_event(msg) != true) {
+            LOG(ERROR) << "User's netlink handler function failed!";
+            return DWPAL_FAILURE;
+        }
+        return DWPAL_SUCCESS;
+    };
+    auto nl_handler_cb = [](struct nl_msg *msg) -> DWPAL_Ret { return nl_handler_cb_wrapper(msg); };
+
+    if (m_fd_nl_events > 0) {
+        //parsing will be done in callback function
+        if (dwpal_driver_nl_msg_get(m_dwpal_nl_ctx, DWPAL_NL_UNSOLICITED_EVENT, NULL,
+                                    nl_handler_cb) == DWPAL_FAILURE) {
+            LOG(ERROR) << " dwpal_driver_nl_msg_get failed,"
+                       << " ctx=" << m_dwpal_nl_ctx;
+            detach();
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool base_wlan_hal_dwpal::dwpal_nl_cmd_set(const std::string &ifname, unsigned int nl_cmd,
+                                           unsigned char *vendor_data, size_t vendor_data_size)
+{
+    bool err = true;
+    if ((vendor_data_size > 0) && (vendor_data == NULL)) {
+        LOG(ERROR) << __FUNCTION__ << "ifname or vendor_data is NULL ==> Abort!\n";
+        return false;
+    }
+
+    if (dwpal_driver_nl_cmd_send(m_dwpal_nl_ctx, DWPAL_NL_UNSOLICITED_EVENT, (char *)ifname.c_str(),
+                                 NL80211_CMD_VENDOR, DWPAL_NETDEV_ID,
+                                 (enum ltq_nl80211_vendor_subcmds)nl_cmd, vendor_data,
+                                 vendor_data_size) != DWPAL_SUCCESS) {
+        err = false;
+    }
+
+    return err;
+}
+
+bool base_wlan_hal_dwpal::dwpal_nl_cmd_get(const std::string &ifname, unsigned int nl_cmd,
+                                           unsigned char *out_buffer, size_t *out_buffer_size,
+                                           const size_t max_buffer_size)
+{
+
+    if ((out_buffer_size == NULL) || (out_buffer == NULL)) {
+        LOG(ERROR) << __FUNCTION__ << "out_buffer or out_buffer_size incorrect ==> Abort!\n";
+        return false;
+    }
+
+    /* Handle a command which invokes an event with the output data */
+    if (dwpal_driver_nl_cmd_send(
+            m_dwpal_nl_ctx, DWPAL_NL_SOLICITED_EVENT, (char *)ifname.c_str(), NL80211_CMD_VENDOR,
+            DWPAL_NETDEV_ID, (enum ltq_nl80211_vendor_subcmds)nl_cmd, NULL, 0) == DWPAL_FAILURE) {
+        LOG(DEBUG) << __FUNCTION__ << "ERROR for cmd = " << nl_cmd;
+        return false;
+    }
+
+    // Passing a lambda with capture is not supported for standard C function
```

Comment:
Can you add a URL which proves this is not supported? for example, the top answer in https://stackoverflow.com/questions/28746744/passing-capturing-lambda-as-function-pointer explains this perfectly with references to the c++11 standard.

Reply:
This pattern with the static capturing lambda variable is repeated a couple of times. AFAIU it is only needed because dwpal doesn't have a reentrant API. So maybe it's better to create a function that does this wrapping and that takes a C++ (instead of C) function as argument.

Reply:
There method also used in base_wlan_hal_nl80211.cpp .
If you want to refactor this code it should be done in different PR.